test/e2e/split/main.go

```diff
@@ -0,0 +1,226 @@
+		}
+		for _, possibleNames := range specNames {
+			if len(possibleNames) != 2 {
+				log.Printf("%s: expected to find 2 submatch, found %d:", match, len(possibleNames))
+				for _, name := range possibleNames {
+					log.Printf("\t%s\n", string(name))
+				}
+				continue
+			}
+			describe := strings.TrimSpace(string(possibleNames[1]))
+			describeTable[describe] = struct{}{}
+		}
+	}
+
+	describes := make([]string, len(describeTable))
+	i := 0
+	for describeKey := range describeTable {
+		describes[i] = describeKey
+		i++
+	}
+	return describes, nil
+}
+
+func createChunkRegexp(numChunks, printChunk int, specs []string) (string, error) {
+
+	if printChunk >= numChunks {
+		return "", fmt.Errorf("the chunk to print (%d) must be a smaller number than the number of chunks (%d)", printChunk, numChunks)
+	}
+
+	numSpecs := len(specs)
+	if numSpecs < numChunks {
+		return "", fmt.Errorf("have more desired chunks (%d) than specs (%d)", numChunks, numSpecs)
+	}
+
+	// Create chunks of size ceil(number of specs/number of chunks) in alphanumeric order.
+	// This is deterministic on inputs.
+	chunks := make([][]string, numChunks)
+	interval := int(math.Ceil(float64(numSpecs) / float64(numChunks)))
+	currIdx := 0
+	for chunkIdx := 0; chunkIdx < numChunks; chunkIdx++ {
+		nextIdx := int(math.Min(float64(currIdx+interval), float64(numSpecs)))
+		chunks[chunkIdx] = specs[currIdx:nextIdx]
+		currIdx = nextIdx
+	}
+
+	chunk := chunks[printChunk]
+	if len(chunk) == 0 {
+		// This is a panic because the caller may skip this error, resulting in missed test specs.
+		panic(fmt.Sprintf("bug: chunk %d has no elements", printChunk))
+	}
+
+	// Write out the regexp to focus chunk specs via `ginkgo -focus <re>`.
+	var reStr string
+	if len(chunk) == 1 {
+		reStr = fmt.Sprintf("%s .*", chunk[0])
+	} else {
+		sb := strings.Builder{}
+		sb.WriteString(chunk[0])
+		for _, test := range chunk[1:] {
+			sb.WriteString("|")
+			sb.WriteString(test)
+		}
+		reStr = fmt.Sprintf("(%s) .*", sb.String())
+	}
+
+	return reStr, nil
+}
+
+func findMinimalWordPrefixes(specs []string) (prefixes []string) {
+
+	// Create a word trie of all spec strings.
+	t := make(wordTrie)
+	for _, spec := range specs {
+		t.push(spec)
+	}
+
+	// Now find the first branch point for each path in the trie by DFS.
+	for word, node := range t {
+		var prefixElements []string
+	next:
+		if word != "" {
+			prefixElements = append(prefixElements, word)
+		}
+		if len(node.children) == 1 {
+			for nextWord, nextNode := range node.children {
+				word, node = nextWord, nextNode
+			}
+			goto next
+		}
+		// TODO: this might need to be joined by "\s+"
+		// in case multiple spaces were used in the spec name.
+		prefixes = append(prefixes, strings.Join(prefixElements, " "))
+	}
+
+	return prefixes
+}
+
+// wordTrie is a trie of word nodes, instead of individual characters.
+type wordTrie map[string]*wordTrieNode
+
+type wordTrieNode struct {
+	word     string
+	children map[string]*wordTrieNode
+}
```

Comment:
Any particular reason you settled on a trie implementation vs. a more primitive implementation of sorting test specs and chunking arbitrarily?

Reply:
Primarily to prevent one spec being run by two different chunks. Imagine running two specs defined by phrases "Foo" and "Foo bar" naively in two chunks: the first chunk (`^Foo.+`) would run both specs, while the second chunk (`^Foo bar.+`) would run only the latter spec. By using a word trie, the phrases are reduced down to just the word "Foo", which will run both specs in one chunk. At first the downside appears to be a less even distribution of chunk runtimes, since "Foo" and "Foo bar" will always be run together; however this downside also affects naive chunking because spec phrase prefixes are still shared.

Reply:
This is probably a really naive question, but, why not just use Ginkgo's parallelization: e.g. `ginkgo -procs=N`

Reply:
`-procs` gives you core parallelism, the actions matrix + chunks gives you core and runner parallelism. Runner parallelism is important here because we cannot choose machine size.

Reply:
It's also important to note that the way our CI environment is setup, we create a kind cluster everytime we spawn a new e2e process, so this PR simply builds on top of that existing setup and chunks test specs and feeds those test specs as focus regex groups that an individual kind cluster will then run. There's likely a cleaner solution going forward, but we'd have to shift around some of our current setup to get it working correctly.